### ISSUE TITLE
Android Build Cache feature breaks Android completion

### ISSUE THREAD

**Issue by DonnieWest** [artur-shaik/vim-javacomplete2#325]
tl;dr - the latest Android Studio changes breaks completion of Javacomplete2 on Android projects. I've a workaround working but I will be investigating if I can get something better working.

## Actual behavior

Using Android Studio 2.3 and the corresponding Android Gradle plugin doesn't allow Javacomplete to complete certain classes in Android projects

## Expected behavior

All completions should be present when completing an Android Project

## The steps to reproduce actual behavior (Required!)
 1. Create new Android project using latest Android Studio
 2. Attempt to complete any class that inherits from App Compat

## Background

In the past, Android projects would unzip AAR files as part of the build process into an "exploded-aars" folder in the project which gave Javacomplete2 the jars necessary to complete Android projects properly. As of Android Studio 2.3 this folder has been moved out to a generic location (generally `~/.android/cache` on unix systems) and is now out of the reach of Javacomplete2

## Solutions

1. Have the user disable the build cache feature by adding the following line to their gradle.properties file

```
android.enableBuildCache=false
```

But this is not sustainable and slows down the user's project build times.

2. Unzip the files within the gradle plugin

We can unzip the contents of any AAR file straight into the gradle project's build directory and include it in this plugin.

This is the solution I have working locally _right now_ but I don't like the workaround. It inflates the user's build directory and simply bypasses the Build Cache feature manually.

3. Use the exploded AARs in the user's Gradle Build Cache directory

This is the ideal, but I don't really have an idea of how to get the exploded AARs for the current project. If I include all the jars from the build cache directory, I'll have conflicting jars and classes included. If anyone has a good idea on how to do this, I'll be open to hearing them :+1:

**Comment by teasp00n** [comment]
Perhaps using symlinks from the project build directory to the gradle build directory (where the exploded aars now are)

There are some potential issues with this however:

1. The user removes javacomplete2 but the symlinks remain
2. The user switches back to an older version of gradle/ android plugin such that the symlinks actually cause unintended side effects (their build files would still go to the gradle location and not project local)
3. Not sure if windows has a 'symlink' equivalent if cross platform is important to the project.

Thoughts?

**Comment by DonnieWest** [comment]
@teasp00n I was thinking along those lines as well. The issue I have right now is that the gradle build cache directory includes the packages for EVERY gradle project built on that machine. This gives us two big issues:

1. The size of the classpath will grow with each new project compiled or each library upgraded and cause JavaComplete2 to slow down and complete things that aren't valid for the current project

2. If two libraries have conflicting versions of a library, both versions will be added to the classpath and cause issues with each project.

So, ideally, I need a way to ONLY pull in the libraries for the current project I just don't see a way to with the current system :/

**Comment by teasp00n** [comment]
@DonnieWest have you gotten any further with this or are you still using your workaround?

**Comment by DonnieWest** [comment]
@teasp00n I've not had the opportunity to look into this further unfortunately

**Comment by DonnieWest** [comment]
@teasp00n and anyone else interested in trying this out - I've gotten a rather naive fix up at #336 and I'd appreciate some feedback

**Comment by IgorGanapolsky** [comment]
Unfortunately, `android.enableBuildCache=false` solution doesn't work with **com.android.tools.build:gradle:3.0.0-beta3**

**Comment by DonnieWest** [comment]
@IgorGanapolsky the latest version of this VIM plugin shouldn't need that. Are you encountering issues with getting completions?

**Comment by IgorGanapolsky** [comment]
@DonnieWest I need to get the exploded AARs in build/intermediates directory (as per your top post).  But they aren't there.

**Comment by DonnieWest** [comment]
@IgorGanapolsky it shouldn't matter - the plugin can currently check the gradle build cache (rather naively) for the files it needs and use those. Do you get proper autocompletion on your project? If not, I'd appreciate looking at it or some other small project that replicates your issue so I can fix it :)
